sbncode/SBNEventWeight/App/SystToolsEventWeight_module.cc

```diff
@@ -0,0 +1,238 @@
+
+  SystToolsEventWeight(SystToolsEventWeight const &) = delete;
+  SystToolsEventWeight(SystToolsEventWeight &&) = delete;
+  SystToolsEventWeight& operator = (SystToolsEventWeight const&) = delete;
+  SystToolsEventWeight& operator = (SystToolsEventWeight&&) = delete;
+
+private:
+  void produce(art::Event& e) override;
+  void beginRun(art::Run& run) override;
+
+private:
+  std::string fGenieModuleLabel;
+  bool fAllowMissingTruth;
+  bool fDebugMode;
+
+  systtools::provider_list_t fSystProviders;
+  systtools::param_header_map_t fParamHeaderMap;
+  systtools::ParamHeaderHelper fParamHeaderHelper;
+
+};
+
+
+SystToolsEventWeight::SystToolsEventWeight(fhicl::ParameterSet const& p)
+  : EDProducer{p}
+{
+
+  fGenieModuleLabel = p.get<std::string>("generator_module_label", "generator");
+  fAllowMissingTruth = p.get<bool>("AllowMissingTruth");
+  fDebugMode = p.get<bool>("debugMode");
+
+  fhicl::ParameterSet syst_provider_config = p.get<fhicl::ParameterSet>("generated_systematic_provider_configuration");
+
+  MF_LOG_INFO("SystToolsEventWeight") << "Configuring ISystProvider";
+
+  fSystProviders = systtools::ConfigureISystProvidersFromParameterHeaders(syst_provider_config);
+  fParamHeaderMap = systtools::BuildParameterHeaders(fSystProviders);
+  fParamHeaderHelper = systtools::ParamHeaderHelper(fParamHeaderMap);
+
+  MF_LOG_INFO("SystToolsEventWeight") << "ISystProvider is configuered" << std::endl;
+
+  produces<std::vector<sbn::evwgh::EventWeightMap> >();
+  produces<art::Assns<simb::MCTruth, sbn::evwgh::EventWeightMap> >();
+  produces<std::vector<sbn::evwgh::EventWeightParameterSet>, art::InRun>();
+
+}
+
+
+void SystToolsEventWeight::produce(art::Event& e) {
+
+  auto mcwghvec = std::make_unique<std::vector<sbn::evwgh::EventWeightMap> >();
+  auto wghassns = std::make_unique<art::Assns<simb::MCTruth, sbn::evwgh::EventWeightMap> >();
+
+  art::PtrMaker<sbn::evwgh::EventWeightMap> makeWeightPtr(e);
+
+  std::vector<art::Ptr<simb::MCTruth> > mclist;
+  art::Handle<std::vector<simb::MCTruth>> mcTruthHandle;
+  if(!fGenieModuleLabel.empty()) e.getByLabel(fGenieModuleLabel, mcTruthHandle);
+  // Prooceed even with missing handle if we want to require the MCTruth to be
+  // found, so that an exception will be thrown explaining the problem.
+  if(mcTruthHandle.isValid() || !fAllowMissingTruth){
+
+    art::fill_ptr_vector(mclist, mcTruthHandle);
+
+    if(fDebugMode){
+      std::cout << "[SystToolsEventWeight::produce] mclist.size() = " << mclist.size() << "\n"
+                << "[SystToolsEventWeight::produce] fSystProviders.size() = " << fSystProviders.size() << std::endl;
+    }
+
+    for( auto &sp : fSystProviders ){
+
+      std::unique_ptr<systtools::EventResponse> syst_resp = sp->GetEventResponse(e);
+      if( !syst_resp ){
+        std::cout << "[ERROR]: Got nullptr systtools::EventResponse from provider "
+                  << sp->GetFullyQualifiedName();
+        throw std::exception();
```

Comment:
Please throw an _art_ exception:
```suggestion
        throw cet::exception{ "SystToolsEventWeight" }
          << "Got nullptr systtools::EventResponse from provider "
          << sp->GetFullyQualifiedName() << "\n";
```

Reply:
Updated as suggested.